ttnn/cpp/ttnn/distributed/distributed_tensor.cpp

```diff
@@ -0,0 +1,192 @@
+        if (!col_dim.has_value()) {
+            for (const auto& t : row_tensors) {
+                for (int i = 0; i < cols; ++i) {
+                    tensor_shards.push_back(t);
+                }
+            }
+        } else {
+            for (const auto& t : row_tensors) {
+                auto col_chunks = experimental::xtensor::chunk(t, cols, *col_dim);
+                tensor_shards.insert(tensor_shards.end(), col_chunks.begin(), col_chunks.end());
+            }
+        }
+
+        TT_FATAL(
+            static_cast<int>(tensor_shards.size()) == rows * cols,
+            "ShardTensor2dMesh: Sharding failed. Number of shards should match the product of the mesh "
+            "dimensions. Size: {}, rows: {}, cols: {}",
+            tensor_shards.size(),
+            rows,
+            cols);
+
+        return tensor_shards;
+    }
+
+    DistributedTensorConfig config() const override { return DistributedTensorConfig{ShardTensor2D(mesh_shape_)}; }
+
+private:
+    MeshShape mesh_shape_;
+    Shard2dConfig config_;
+};
+
+class ConcatMeshToTensor : public MeshToTensor {
+public:
+    ConcatMeshToTensor(int dim) : concat_dim_(dim) {}
+
+    Tensor compose(const std::vector<Tensor>& tensors) override {
+        return experimental::xtensor::concatenate(tensors, concat_dim_);
+    }
+
+private:
+    int concat_dim_ = -1;
+};
+
+class ConcatMesh2dToTensor : public MeshToTensor {
+public:
+    ConcatMesh2dToTensor(MeshDevice& mesh_device, const Concat2dConfig& config) :
+        mesh_shape_(mesh_device.shape()), config_(config) {}
+
+    Tensor compose(const std::vector<Tensor>& tensors) override {
+        const auto [rows, cols] = mesh_shape_;
+        const auto [row_dim, col_dim] = config_;
+
+        std::vector<Tensor> row_concatenated;
+        row_concatenated.reserve(rows);
+        for (int i = 0; i < rows; ++i) {
+            auto row_start = tensors.begin() + i * cols;
+            auto row_end = row_start + cols;
+            std::vector<Tensor> row_tensors(row_start, row_end);
+            row_concatenated.push_back(experimental::xtensor::concatenate(row_tensors, col_dim));
+        }
+
+        return experimental::xtensor::concatenate(row_concatenated, row_dim);
+    }
+
+private:
+    MeshShape mesh_shape_;
+    Concat2dConfig config_;
+};
+
+}  // namespace
+
+std::unique_ptr<TensorToMesh> replicate_tensor_to_mesh_mapper(MeshDevice& mesh_device) {
+    return std::make_unique<ReplicateTensorToMesh>(mesh_device.num_devices());
+}
+
+std::unique_ptr<TensorToMesh> shard_tensor_to_mesh_mapper(MeshDevice& mesh_device, int dim) {
+    return std::make_unique<ShardTensorToMesh>(mesh_device.num_devices(), dim);
+}
+
+std::unique_ptr<TensorToMesh> shard_tensor_2d_to_mesh_mapper(
+    MeshDevice& mesh_device, const MeshShape& mesh_shape, const Shard2dConfig& config) {
+    TT_FATAL(
+        config.row_dim.has_value() || config.col_dim.has_value(),
+        "ShardTensor2dMesh requires at least one dimension to shard");
```

Comment:
Style inconsistency `ShardTensor2dMesh` here, `ShardTensor2dMesh:` below.

Reply:
Fixed the inconsistency - referring to '2d mesh' everywhere in class names and the creation functions.